docs/dev/architecture.rst

```diff
@@ -0,0 +1,127 @@
+   from Normandy and executing them within Firefox.
+
+.. note:: Normandy provides both the names of actions and, separately,
+   implmentations for them to be used in the self-repair system addon. Other
+   Firefox features may also query Normandy and implement the actions however
+   they wish, and are not required to use the implementations that Normandy
+   provides.
+
+Normandy / Self-Repair Protocol
+-------------------------------
+The following section describes the protocol used to communicate between
+Normandy and the Self-Repair system addon. The domain ``shield.mozilla.org``
+is used as a stand-in domain; the actual domain Normandy is hosted at may be
+different.
+
+.. warning:: The following documentation is still in flux and does not describe
+   a live system yet. Expect the protocol to change.
+
+Retrieving Recipes
+^^^^^^^^^^^^^^^^^^
+The Self-Repair addon retrieves a list of recipes by making an HTTP POST request
+to the URL::
+
+   https://shield.mozilla.org/api/v1/fetch_bundle
+
+The request should contain a JSON object of the form:
+
+.. code-block:: json
+
+   {
+      "locale": "en-US",
+      "firefox_version": "42.0.1"
+   }
+
+The API will respond with a JSON object:
+
+.. code-block:: json
+
+   {
+      "recipes": [
+         {
+            "name": "heartbeat.australis.1",
+            "actions": [
+               {
+                  "name": "show_survey",
+                  "impl_hash": "4011cf569f3639b9069fc2c50117d3f8597e83c61be2014081ad06ee0fe427c2",
+                  "arguments": {
+                     "url": "https://somesurvey.com/blah"
+                  }
+               }
+            ]
+         },
+         {
+            "name": "remove.bad.toolbar",
+            "actions": [
+               {
+                  "name": "check_for_bad_toolbar",
+                  "impl_hash": "4011cf569f3639b9069fc2c50117d3f8597e83c61be2014081ad06ee0fe427c2",
+                  "arguments": {
+                     "toolbar_name": "Shady Tim's Totally Legit Search"
+                  }
+               },
+               {
+                  "name": "telemetry_ping",
+                  "impl_hash": "4011cf569f3639b9069fc2c50117d3f8597e83c61be2014081ad06ee0fe427c2",
+                  "arguments": {
+                     "id": "shady_tim_strikes_again"
+                  }
+               }
+            ]
+         }
+      ]
+   }
+
+Retrieving an Action
+^^^^^^^^^^^^^^^^^^^^
+The Self-Repair addon retrieves the code necessary to execute an action by
+making an HTTP GET to the URL::
+
+   https://shield.mozilla.org/api/v1/get_action_impl?name=action_name
+
+The API will respond with a JSON object of the form:
+
+.. code-block:: json
+
+   {
+      "impl": "function action(selfRepairAPI, args) { /* Do something */ }"
+   }
+
+Legacy Self-Repair
+------------------
+Normandy also hosts an endpoint so that it can replace the predecessor
+Self-Repair server, which delivers a single JavaScript file for Firefox to
```

Comment:
The old Self-Repair server serves a single HTML file that contains inline JS. We will have to do the same. (We might be able to not make it inline, but I don't see what that would gain us.)